### ISSUE TITLE
Unicode, Emojis, and Wonkiness Oh My.

### ISSUE THREAD

**Issue by jared-w** [matterhorn-chat/matterhorn#685]
tl;dr Matterhorn inherits some apparently wonkiness from Haskell (somewhere) and that makes using unicode with it a bit messy at times.

If I may preach to the choir for a brief second, I do think it's important to remember that, with terminals + unicode, a large part of the brokenness depends on one's font, terminal, and more; all matterhorn can reasonably be expected to do is "as best effort as possible" given the constraints of its environment. As such, it's quite possible that some of these examples will never work "correctly" (because correct is not really an unambiguous or reasonable criteria here).That said, I have taken care in my setup to eliminate most of those limitations, and I include an example of "best case working" (which, while not perfect, is pretty solid).

(N.B.: if it should ever actually matter, nitty gritty details into my OS+env stack can be found in my [dotfiles](https://github.com/jared-w/nixos-configs) which are fully reproducible and hermetic.)

First, some background. Below are a list of emojis that will be used as running examples of "it doesn't work great". One of which is not pasteable into matterhorn at all, and the rest which are increasingly "broken".

```
  -- The "docker" emoji if one is using a patched nerd font (example of a private use area emoji symbol). Note: unpastable in GHCi
🙏 -- pray-hands emoji. A double width emoji that renders "correctly"
🤦🏾‍♂️ -- man facepalm emoji
👨 👩 👧 -- the below emoji split up without zero-width-joiners
👨‍👩‍👧.
```

Here's a [link](https://eclecticlight.co/2018/03/15/compound-emoji-can-confuse/) to an article which lists some of the counter-examples.

![Screenshot from 2021-01-12 18-06-14](https://user-images.githubusercontent.com/27441417/104397255-45d75400-5501-11eb-9298-05ededfcb016.png)

This first screenshot is the result of pasting each line one by one into a text chat in matterhorn. Note, the first line with the docker emoji did not paste at all (and GHCi will refuse to let it be pasted as well, for reference)

![Screenshot from 2021-01-12 18-16-09](https://user-images.githubusercontent.com/27441417/104397754-48867900-5502-11eb-8ad1-fdb10d28d9ec.png)

This is what these emojis look like in my terminal+neovim currently. So, support isn't perfect (ZWJs in particular are a little wild for TUI text editors; they make a lot of strong assumptions about text, characters, width, etc., that isn't holding up well in the modern unicode era and never really worked well for scripts such as arabic in the first place.)

![Screenshot from 2021-01-12 18-12-11](https://user-images.githubusercontent.com/27441417/104397488-c4cc8c80-5501-11eb-811b-fa0b9597b72e.png)

This is what it looks like in my raw terminal (using [bat](https://github.com/sharkdp/bat) for cat, zsh, a patched nerd font, and a few other goodies. This level of emoji support is, as far as I know, best case theoretical support for a terminal. (You can see how the family emoji renders "correctly" but still takes up 2 extra spaces of width and has a very small period at the end of the "width" of the emoji.

---

Now for an example of wonky UI.

Assuming you paste any pasteable emoji that is "wide" into the matterhorn chat and then continue to type, it'll look something like this:

![Screenshot from 2021-01-12 17-54-20](https://user-images.githubusercontent.com/27441417/104398124-03af1200-5503-11eb-8fd0-56b5e5af16e4.png)

---

Lastly, here's a text snippet showing how Data.Text mangles unicode a bit when `show`ing it as a string.

![Screenshot from 2021-01-12 17-53-29](https://user-images.githubusercontent.com/27441417/104397950-ae730080-5502-11eb-9fcf-761e612e00f6.png)

(The 'T' in 'Tprintf' does not exist, it is really `printf` from `Text.Printf`, and is a visual artifact left over from pressing the up arrow to go up in history. Actually pasting those in as expressions without clearing the terminal screen and forcing a redraw leads to much worse visual glitches)

This is important because, currently, this is the sort of conversion you get when trying to render any messages with emojis using a notification script. Text.Printf doesn't seem to do this and lets even the private use unicode stuff go through correctly.

**Comment by jtdaugherty** [comment]
I know we discussed this elsewhere, but to be explicit here: it would be helpful if this could be split up to report one problem at a time with a minimal test case for each. In addition, if one of these issues is about Unicode character widths being wrong, that's already a known problem and doesn't need to be reported again (and is intractable to fix).

**Comment by jtdaugherty** [comment]
I know @jared-w hasn't gotten around to breaking this down a bit more, but since I want to avoid having this linger, I want to comment on some of the cases mentioned.

First I want to comment on the problematic emoji cases. Due to some of the issues that Jared cited, the terminal emulator in use will disagree with Matterhorn about the width of some emoji. The Vty library, which is responsible for Matterhorn's terminal I/O, had some work done on it to try to query the terminal about the widths of all Unicode characters. That helped somewhat, but 1) Matterhorn has not had that work integrated into it yet because it is not working well enough and 2) that work only helps for Unicode characters that can be encoded as a single code point. Both the Vty and Brick packages, which we use together to run Matterhorn's UI, do not support width computation of multi-code-point characters (such as ones involving zero-width joiners). As much as it pains me to say this, I really don't have any idea right when, if ever, that will be supported. (Contributions welcome, but the design issues alone need to be sorted out first.) Even GHCI's terminal I/O processing cannot handle the "praying hands" emoji, as Jared mentioned above.

Second, on the topic of entering characters into the editor, the lack of support for widths of various Unicode characters extends to the Brick editor as well. That's why the screenshot of editing above shows some rendering issues; the width of the input characters is underestimated, leading to overdrawing in the editor's screen region.

Lastly, I am not sure what the `printf` example is intended to show. The result of `T.pack "🙏"` above is exactly what I would expect; it's printing the representation of the string, not the string itself. Using `Data.Text.IO.putStrLn` on the resulting text value prints what I would expect, too, just as `printf` did above:
```
Prelude Data.Text Data.Text.IO> Data.Text.IO.putStrLn (pack "🙏")
🙏
```

**Comment by jtdaugherty** [comment]
@jared-w I'd love to keep discussing this in the form of more specific tickets. I'm going to close this since it's unclear to me what we can do to move forward on it. Please open new tickets if you like, and we can continue the discussion there. Thanks!
